src/lib.rs

```diff
@@ -166,79 +155,94 @@ impl Client {
     /// beat task (if running). If we still have a connection to faktory, this will
     /// send the END command as well.
     pub fn close(&self) -> Result<()> {
-        self.shutdown_sender.send(())?;
+        // If the receiver isn't there, there is nothing to close
+        let _ = self.shutdown_sender.send(());
         Ok(())
     }
 
-    pub async fn send_command(&self, command: FaktoryCommand) -> Result<FaktoryResponse> {
+    async fn send_command(&self, command: FaktoryCommand) -> Result<FaktoryResponse> {
         let (response_sender, response_receiver) = oneshot::channel();
 
-        self.command_sender.send((command, response_sender)).await?;
+        self.command_sender
+            .send((command, response_sender))
+            .await
+            .map_err(|_| Error::SendCommand)?;
         match response_receiver.await? {
-            FaktoryResponse::Error(err) => Err(err),
+            FaktoryResponse::Error(kind, msg) => Err(Error::Faktory(kind, msg)),
             other => Ok(other),
         }
     }
 
-    async fn handle_msg(msg: FaktoryCommandMessage, connection: &mut Connection) -> CommandOutcome {
+    async fn handle_msg(
+        msg: FaktoryCommandMessage,
+        connection: &mut Connection,
+        config: &Config,
+        shutdown_channel: &broadcast::Receiver<()>,
+    ) {
         let (cmd, responder) = msg;
 
-        match Client::handle_cmd(cmd, connection).await {
-            FaktoryResponse::Error(err) => match err {
-                Error::FaktoryLib(faktory_lib_async::Error::Io(_)) => {
-                    responder.send(FaktoryResponse::Retry).unwrap();
-                    CommandOutcome::Reconnect
+        loop {
+            match Client::handle_cmd(&cmd, connection).await {
+                // If the response receiver has gone away, it's because the caller killed it's tokio task
+                // That means faktory will still get the command, but the caller won't get the response
+                // So we can safely ignore channel failures here
+
+                Err(faktory_lib_async::Error::ReceivedErrorMessage(kind, msg)) => {
+                    let _ = responder.send(FaktoryResponse::Error(kind, msg));
+                    break;
                 }
-                _ => {
-                    responder.send(FaktoryResponse::Error(err)).unwrap();
-                    CommandOutcome::Ok
+                Ok(ok_response) => {
+                    let _ = responder.send(ok_response);
+                    break;
+                }
+
+
+                Err(_err) => {
+                    // TODO: is there a case where retrying will keep failing because the user provided data
+                    // for this request caused an unexpected error? should we give up, or re-enqueue this
+                    // command at the end of the line?
+
+                    //error!("faktory_lib_async returned a critical error, reconnecting and trying to send the message again: {_err}");
```

Comment:
My reading of the protocol is that if we get ERR back most of the time we should *not* retry, since it means malformed input and we'll just get stuck in a retry loop.

There are probably exceptions. Looking at the faktory server code, I think the biggest one would be if we get ERR back because the internal Redis server had problems.